### ISSUE TITLE
Date parsed as String with inferSchema =true

### ISSUE THREAD

**Issue by suhasDataken** [databricks/spark-csv#170]
I edited the file comments.csv (/src/test/resources/comments.csv) from my local clone of spark-csv and  deleted the time part of time-stamp fields such that we only have dates now as shown below:
~ Version 1.0
~ Using a non-standard comment char to test CSV parser defaults are overridden
1,2,3,4,5.01,2015-08-20
6,7,8,9,0,2015-08-21
~0,9,8,7,6,2015-08-22
1,2,3,4,5,2015-08-23

And got the following result (date parsed as string):
~ Version 1.0
~ Using a non-standard comment char to test CSV parser defaults are overridden
1,2,3,4,5.01,2015-08-20
6,7,8,9,0,2015-08-21
~0,9,8,7,6,2015-08-22
1,2,3,4,5,2015-08-23

With the following Java code:
Map<String, String> paramMap = new HashMap<String, String>();
		paramMap.put("header", "true");
		paramMap.put("inferSchema", "true");
		paramMap.put("comment", "~");
		paramMap.put("path", "D:\\Path\\Repository\\comments.csv");		
		DataFrame df = sqlContext.read().format("com.databricks.spark.csv").options(paramMap).load();
		df.printSchema();

**Comment by falaki** [comment]
What version of spark-csv are you using?

**Comment by suhasDataken** [comment]
I've built the jar from the latest from master.

**Comment by ImNotMyself** [comment]
In InferSchema.scala, it is only trying to parse Timestamp but no Date type.
To support parsing Date, you may need to modify the InferSchema. 
Below is my solution:
```scala
  import java.sql.Date //add import

  private[csv] def inferField(typeSoFar: DataType, field: String): DataType = {
    if (field == null || field.isEmpty) {
      typeSoFar
    } else {
      typeSoFar match {
        case NullType => tryParseInteger(field)
        case IntegerType => tryParseInteger(field)
        case LongType => tryParseLong(field)
        case DoubleType => tryParseDouble(field)
        case TimestampType => tryParseTimestamp(field)
        case DateType => tryParseDate(field)    // Add this try
        case StringType => StringType
        case other: DataType =>
          throw new UnsupportedOperationException(s"Unexpected data type $other")
      }
    }
  }

  def tryParseTimestamp(field: String): DataType = {
    if ((allCatch opt Timestamp.valueOf(field)).isDefined) {
      TimestampType
    } else {
      tryParseDate(field)  //  pass to tryParseDate
    }
  }

  def tryParseDate(field: String): DataType = {
    if ((allCatch opt Date.valueOf(field)).isDefined) {
      DateType
    } else {
      stringType()
    }
  }
```

@falaki will such Date parsing be implemented to InferSchema officially ? any plan ?

**Comment by falaki** [comment]
being able to support arbitrary date and time formats, possibly through an optional parameter was planned. I was hoping the community will submit patches for those improvements :-)

**Comment by cdelgadob** [comment]
@falaki : I think an optional parameter is the right way. One question: Should we assume that all the Date fields share the same format, or should we allow to define more than one?
I personally think all Date fields with the same format, otherwise it gets messy and possibly slower. What do you think?

**Comment by falaki** [comment]
Yes, I agree. For now we can assume all date fields follow the same format.

**Comment by sasidhar2207** [comment]
Hi , i am trying to infer schema from a CSV file, date column is taking as String, How can i infer date datatype with the date format in column.
As we have so many date formats. task to infer dateformat  as per in file column instead of giving manually everytime.

Could anyone help me on this 

i have sample code, which may be helpful.

package scala_programs

import java.text.ParseException

import java.text.SimpleDateFormat

object DateFormatParser {

  private val formats: Array[String] = Array(
    "yyyy-MM-dd'T'HH:mm:ss'Z'",
    "yyyy-MM-dd'T'HH:mm:ssZ",
    "yyyy-MM-dd'T'HH:mm:ss",
    "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
    "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
    "yyyy-MM-dd HH:mm:ss",
    "MM/dd/yyyy HH:mm:ss",
    "MM/dd/yyyy'T'HH:mm:ss.SSS'Z'",
    "MM/dd/yyyy'T'HH:mm:ss.SSSZ",
    "MM/dd/yyyy'T'HH:mm:ss.SSS",
    "MM/dd/yyyy'T'HH:mm:ssZ",
    "MM/dd/yyyy'T'HH:mm:ss",
    "yyyy:MM:dd HH:mm:ss",
    "yyyyddMM",
    "dd-MM-yyyy",
    "yyyy-MM-dd",
    "dd-MM-yyyy",
  "dd.MM.yyyy",
  "dd/MM/yyyy",
  "dd:MM:yyyy",
  "MM-dd-yyyy",
  "MM.dd.yyyy",
  "MM/dd/yyyy",
  "MM:dd:yyyy",
  "yyyy-MM-dd",
  "yyyy.MM.dd",
  "yyyy:MM:dd",
  "yyyy/MM/dd",
  "yyyyddMM",
  "ddMMyyyy",
  "yyyyMMdd"
  )

  def main(args: Array[String]): Unit = {
    val yyyyMMdd: String = "12-09-2017"
    parse(yyyyMMdd)
  }

  def parse(d: String): Unit = {
    if (d != null) {
      for (parse <- formats) {
        val sdf: SimpleDateFormat = new SimpleDateFormat(parse)
        try {

          sdf.parse(d)
          println("Date formats" +parse)
        } catch {
          case e: ParseException => {
          }

        }
      }
    }
  }

}
